site-builder/server/server.ts

```diff
@@ -0,0 +1,586 @@
+			} catch (err) {
+				console.error("Error calculating directory size:", err);
+				// Continue anyway, we'll just have less precise progress reporting
+			}
+		})();
+
+		// Pipe archive output to the pass-through stream
+		archive.pipe(passThrough);
+
+		console.log({ msg: `Creating S3 upload`, bucket, key, passThrough, client });
+		// Configure S3 upload using the pass-through stream as input
+		const upload = new Upload({
+			client,
+			params: {
+				Bucket: bucket,
+				Key: key,
+				Body: passThrough, // Use the pass-through stream instead of archiver directly
+				ContentType: "application/zip",
+			},
+			queueSize: 4,
+			partSize: 1024 * 1024 * 5, // 5MB parts
+			leavePartsOnError: false,
+		});
+
+		// Progress tracking for the upload
+		let uploadLastPercentage = 0;
+		upload.on("httpUploadProgress", (progress) => {
+			if (progress.loaded && progress.total) {
+				const percentage = Math.round((progress.loaded / progress.total) * 100);
+				if (percentage >= uploadLastPercentage + 5 || percentage === 100) {
+					console.log(
+						`Upload progress: ${percentage}% | ${formatBytes(progress.loaded)} of ${formatBytes(progress.total)}`
+					);
+					uploadLastPercentage = percentage;
+				}
+			}
+		});
+
+		// Progress reporting during the compression process
+		archive.on("entry", (entry) => {
+			if (entry.stats && entry.stats.size) {
+				processedBytes += entry.stats.size;
+				if (totalBytes > 0) {
+					const percentage = Math.round((processedBytes / totalBytes) * 100);
+					if (percentage >= lastPercentage + 5 || percentage === 100) {
+						console.log(
+							`Compression progress: ${percentage}% | ${formatBytes(processedBytes)} of ${formatBytes(totalBytes)}`
+						);
+						lastPercentage = percentage;
+					}
+				} else {
+					// If we don't have total size, just report processed bytes
+					console.log(`Compression progress: ${formatBytes(processedBytes)} processed`);
+				}
+			}
+		});
+
+		// Handle warnings from archiver
+		archive.on("warning", (err: ArchiverError) => {
+			if (err.code === "ENOENT") {
+				// log warning
+				console.warn(err);
+			} else {
+				// throw error
+				reject(err);
+			}
+		});
+
+		// Handle errors from archiver
+		archive.on("error", (err: Error) => {
+			console.error("Archive error:", err);
+			reject(err);
+		});
+
+		// Handle errors from the pass-through stream
+		passThrough.on("error", (err) => {
+			console.error("Stream error:", err);
+			reject(err);
+		});
+
+		// Start the upload process
+		upload
+			.done()
+			.then((result) => {
+				console.log(`Archive upload completed: ${fileName}`);
+				resolve(result.Location!);
+			})
+			.catch((err) => {
+				console.error("Upload error:", err);
+				reject(err);
+			});
+
+		// Append files from the directory to the archive
+		archive.directory(sourceDir, false);
+
+		// Finalize the archive - this is when data actually starts flowing
+		archive.finalize();
+	});
+};
```

Comment:
yum streams. im glad i can let robots write this code for me now, i really hate working with this